lib/linkerd2-proxy-discover/src/from_resolve.rs

```diff
@@ -0,0 +1,115 @@
+use futures::{try_ready, Async, Future, Poll};
+use indexmap::IndexSet;
+use linkerd2_proxy_core::resolve::{Resolution, Resolve, Update};
+use std::net::SocketAddr;
+use tower::discover::Change;
+
+#[derive(Clone, Debug)]
+pub struct FromResolve<R> {
+    resolve: R,
+}
+
+#[derive(Debug)]
+pub struct DiscoverFuture<F> {
+    future: F,
+}
+
+/// Observes an `R`-typed resolution stream, using an `M`-typed endpoint stack to
+/// build a service for each endpoint.
+pub struct Discover<R: Resolution> {
+    resolution: R,
+    active: IndexSet<SocketAddr>,
+    pending: Vec<Change<SocketAddr, R::Endpoint>>,
+}
+
+// === impl FromResolve ===
+
+impl<R> FromResolve<R> {
+    pub fn new<T>(resolve: R) -> Self
+    where
+        R: Resolve<T>,
+    {
+        Self { resolve }
+    }
+}
+
+impl<T, R> tower::Service<T> for FromResolve<R>
+where
+    R: Resolve<T> + Clone,
+{
+    type Response = Discover<R::Resolution>;
+    type Error = R::Error;
+    type Future = DiscoverFuture<R::Future>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.resolve.poll_ready()
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        Self::Future {
+            future: self.resolve.resolve(target),
+        }
+    }
+}
+
+// === impl DiscoverFuture ===
+
+impl<F> Future for DiscoverFuture<F>
+where
+    F: Future,
+    F::Item: Resolution,
+{
+    type Item = Discover<F::Item>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let resolution = try_ready!(self.future.poll());
+        Ok(Async::Ready(Discover::new(resolution)))
+    }
+}
+
+// === impl Discover ===
+
+impl<R: Resolution> Discover<R> {
+    pub fn new(resolution: R) -> Self {
+        Self {
+            resolution,
+            active: IndexSet::default(),
+            pending: vec![],
+        }
+    }
+}
+
+impl<R: Resolution> tower::discover::Discover for Discover<R> {
+    type Key = SocketAddr;
+    type Service = R::Endpoint;
+    type Error = R::Error;
+
+    fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::Error> {
+        loop {
+            if let Some(change) = self.pending.pop() {
```

Comment:
I notice that `pending` is essentially being treated as a stack here. Is that actually the correct behavior? 

If I understand correctly, it means that if we buffer a bunch of changes in `pending` and then get polled a bunch of times, we'll see the most recent changes _before_ earlier changes. If the `Discover` is then polled enough times to drain `pending`, then the most recently observed state will be the _oldest_ change.

Should we use a queue/dequeue rather than a stack here?

Reply:
yeah, should be a vecdeque